xarray/core/dataset.py

```diff
@@ -1433,11 +1449,68 @@ def __setitem__(self, key: Hashable, value) -> None:
         variable.
         """
         if utils.is_dict_like(key):
-            raise NotImplementedError(
-                "cannot yet use a dictionary as a key to set Dataset values"
-            )
+            from .dataarray import DataArray
 
-        self.update({key: value})
+            # check for consistency first
+            if isinstance(value, Dataset):
+                missing_vars = [
+                    str(name) for name in value.data_vars if name not in self.data_vars
+                ]
+                if len(missing_vars) > 0:
+                    raise KeyError(
+                        "Variables {} in new values not available in dataset!".format(
+                            ",".join(missing_vars)
+                        )
+                    )
+            for name, var in self.items():
+                missing_keys = [k for k in key.keys() if k not in var.dims]
+                if len(missing_keys) > 0:
+                    raise KeyError(
+                        "Variable {} does not contain dimensions {}!".format(
+                            name, ",".join(missing_keys)
+                        )
+                    )
+
+                # test indexing
+                try:
+                    var_k = var[key]
+                except IndexError:
+                    raise IndexError(
+                        "Indexer {} not available in variable '{}'".format(key, name)
+                    )
```

Comment:
```suggestion
                except IndexError as e:
                    raise IndexError(
                        "Indexer {} not available in variable '{}'".format(key, name)
                    ) from e
```

This will give a better message with the traceback. Instead of `During handling of the above exception, another exception occurred:`, the user will see `The above exception was the direct cause of the following exception:`